### ISSUE TITLE
Standardising matrix arrays (adopt Array API spec)

### ISSUE THREAD

**Issue by KennethEnevoldsen** [embeddings-benchmark/mteb#941]
Currently MTEB does not consistently use a ground truth for vector manipulation. A solution would be to adopt the [Array API spec](https://data-apis.org/array-api/latest/index.html). A potentially simpler solution is to use either torch or numpy.

- The benefit of using numpy is that it is very stable and rarely leads to OOM and related GPU issues.
- The benefit of using PyTorch is that it potentially allows evaluation using GPUs (currently, sklearn does not support this, but might in the future, when they adopt they API spec)

partly related to the discussion in: https://github.com/embeddings-benchmark/mteb/issues/883

This is not a pressing concern atm.

**Comment by Muennighoff** [comment]
Good point; I think using numpy would also allow us to make torch an optional dependency which could be nice as I assume some models do not require it so people have to install it just for mteb

**Comment by KennethEnevoldsen** [comment]
#2234 and #1791 both improve tensor compatibility allowing models to output torch tensors and where possible also used operations that work for both numpy and torch arrays.

This is a list of some of the current outstanding issues:

1) Unsqueeze not in array API spec, we should use expand_dims, but torch does not implement this (https://github.com/pytorch/pytorch/issues/56774). We can use reshape instead.
2) What to do with torch.functional? We would have to reimplent these to make it work. I would be afraid that this could introduce unintended bugs. 

I suspect that these change will be easier once the array spec is more widely adopted.

**Comment by KennethEnevoldsen** [comment]
I found a few places where we could probably remove the `np.asarray` conversion (e.g. in PairClassification)
